From 4a7e71bc0108482c8f22478d70e9e57cc05a60da Mon Sep 17 00:00:00 2001
From: Vaishnav Achath <vaishnav.a@ti.com>
Date: Sat, 8 Feb 2025 18:15:42 +0530
Subject: [PATCH 32/68] TI: arm64: dts: ti: Add J7200 MCSPI loopback overlay

J7200 has main_mcspi4 internally connected as slave to mcu_spi2 master,
this commit adds an overlay for demonstrating the internal MCSPI loopback
feature and acts a reference for enabling DMA and slave mode for MCSPI.
MCSPI slave mode is supported only with DMA enabled. The same overlay can
be used for other J7 platforms (J721E, J721S2, J784S4) with update in the
PSIL thread ID, the necessary information will be added in documentation.

Cc: Ankit Chavda <a-chavda@ti.com>
Signed-off-by: Vaishnav Achath <vaishnav.a@ti.com>
Signed-off-by: Udit Kumar <u-kumar1@ti.com>
---
 arch/arm64/boot/dts/ti/Makefile               | 16 ++++++
 .../dts/ti/k3-j7200-evm-mcspi-loopback.dtso   | 49 +++++++++++++++++++
 2 files changed, 65 insertions(+)
 create mode 100644 arch/arm64/boot/dts/ti/k3-j7200-evm-mcspi-loopback.dtso

diff --git a/arch/arm64/boot/dts/ti/Makefile b/arch/arm64/boot/dts/ti/Makefile
index ce036335b50e..ff8fbdd027ce 100644
--- a/arch/arm64/boot/dts/ti/Makefile
+++ b/arch/arm64/boot/dts/ti/Makefile
@@ -107,6 +107,7 @@ k3-j7200-evm-dtbs := k3-j7200-common-proc-board.dtb k3-j7200-evm-quad-port-eth-e
 dtb-$(CONFIG_ARCH_K3) += k3-j7200-evm.dtb
 dtb-$(CONFIG_ARCH_K3) += k3-j7200-evm-ethfw.dtbo
 dtb-$(CONFIG_ARCH_K3) += k3-j7200-evm-pcie1-ep.dtbo
+dtb-$(CONFIG_ARCH_K3) += k3-j7200-evm-mcspi-loopback.dtbo
 
 # Boards with J721e SoC
 k3-j721e-evm-dtbs := k3-j721e-common-proc-board.dtb k3-j721e-evm-quad-port-eth-exp.dtbo
@@ -244,6 +245,8 @@ k3-am69-sk-pcie0-ep-dtbs := k3-am69-sk.dtb \
 	k3-am69-sk-pcie0-ep.dtbo
 k3-j7200-evm-ethfw-dtbs := k3-j7200-common-proc-board.dtb \
 	k3-j7200-evm-ethfw.dtbo
+k3-j7200-evm-mcspi-loopback-dtbs := k3-j7200-common-proc-board.dtb \
+	k3-j7200-evm-mcspi-loopback.dtbo
 k3-j7200-evm-pcie1-ep-dtbs := k3-j7200-common-proc-board.dtb \
 	k3-j7200-evm-pcie1-ep.dtbo
 k3-j721e-beagleboneai64-dsi-rpi-7inch-panel-dtbs := k3-j721e-beagleboneai64.dtb \
@@ -254,6 +257,8 @@ k3-j721e-common-proc-board-infotainment-dtbs := k3-j721e-common-proc-board.dtb \
 	k3-j721e-common-proc-board-infotainment.dtbo
 k3-j721e-evm-ethfw-dtbs := k3-j721e-common-proc-board.dtb \
 	k3-j721e-evm-ethfw.dtbo
+k3-j721e-evm-mcspi-loopback-dtbs := k3-j721e-common-proc-board.dtb \
+	k3-j7200-evm-mcspi-loopback.dtbo
 k3-j721e-evm-pcie-ntb-dtbs := k3-j721e-common-proc-board.dtb \
 	k3-j721e-evm-pcie-ntb.dtbo
 k3-j721e-evm-pcie0-ep-dtbs := k3-j721e-common-proc-board.dtb \
@@ -264,12 +269,18 @@ k3-j721e-sk-csi2-dual-imx219-dtbs := k3-j721e-sk.dtb \
 	k3-j721e-sk-csi2-dual-imx219.dtbo
 k3-j721s2-evm-pcie1-ep-dtbs := k3-j721s2-common-proc-board.dtb \
 	k3-j721s2-evm-pcie1-ep.dtbo
+k3-j721s2-evm-mcspi-loopback-dtbs := k3-j721s2-common-proc-board.dtb \
+	k3-j7200-evm-mcspi-loopback.dtbo
 k3-j722s-evm-dsi-rpi-7inch-panel-dtbs := k3-j722s-evm.dtb \
 	k3-j722s-evm-dsi-rpi-7inch-panel.dtbo
 k3-j722s-evm-microtips-mf101hie-panel-dtbs := k3-j722s-evm.dtb \
 	k3-j722s-evm-microtips-mf101hie-panel.dtbo
 k3-j742s2-evm-ethfw-dtbs := k3-j742s2-evm.dtb \
 	k3-j742s2-evm-ethfw.dtbo
+k3-j742s2-evm-mcspi-loopback-dtbs := k3-j742s2-evm.dtb \
+	k3-j7200-evm-mcspi-loopback.dtbo
+k3-j784s4-evm-mcspi-loopback-dtbs := k3-j784s4-evm.dtb \
+	k3-j7200-evm-mcspi-loopback.dtbo
 k3-j742s2-evm-pcie-ntb-dtbs := k3-j742s2-evm.dtb \
 	k3-j784s4-j742s2-evm-pcie-ntb.dtbo
 k3-j742s2-evm-usb0-type-a-dtbs := k3-j742s2-evm.dtb \
@@ -314,23 +325,28 @@ dtb- += k3-am625-beagleplay-csi2-ov5640.dtb \
 	k3-am69-sk-csi2-dual-imx219.dtb \
 	k3-am69-sk-pcie0-ep.dtb \
 	k3-j7200-evm-ethfw.dtb \
+	k3-j7200-evm-mcspi-loopback.dtb \
 	k3-j7200-evm-pcie1-ep.dtb \
 	k3-j721e-beagleboneai64-dsi-rpi-7inch-panel.dtb \
 	k3-j721e-beagleboneai64-microtips-mf070zima-lcd3.dtb \
 	k3-j721e-common-proc-board-infotainment.dtb \
 	k3-j721e-evm-ethfw.dtb \
+	k3-j721e-evm-mcspi-loopback.dtb \
 	k3-j721e-evm-pcie-ntb.dtb \
 	k3-j721e-evm-pcie0-ep.dtb \
 	k3-j721e-evm-pcie1-ep.dtb \
 	k3-j721e-sk-csi2-dual-imx219.dtb \
 	k3-j721s2-evm-pcie1-ep.dtb \
+	k3-j721s2-evm-mcspi-loopback.dtb \
 	k3-j722s-evm-dsi-rpi-7inch-panel.dtb \
 	k3-j722s-evm-microtips-mf101hie-panel.dtb \
 	k3-j742s2-evm-ethfw.dtb \
 	k3-j742s2-evm-pcie-ntb.dtb \
 	k3-j742s2-evm-usb0-type-a.dtb \
+	k3-j742s2-evm-mcspi-loopback.dtb \
 	k3-j784s4-evm-ethfw.dtb \
 	k3-j784s4-evm-pcie-ntb.dtb \
+	k3-j784s4-evm-mcspi-loopback.dtb \
 	k3-j784s4-evm-pcie0-pcie1-ep.dtb \
 	k3-j784s4-evm-quad-port-eth-exp1.dtb \
 	k3-j784s4-evm-usb0-type-a.dtb \
diff --git a/arch/arm64/boot/dts/ti/k3-j7200-evm-mcspi-loopback.dtso b/arch/arm64/boot/dts/ti/k3-j7200-evm-mcspi-loopback.dtso
new file mode 100644
index 000000000000..77a02cdbbe36
--- /dev/null
+++ b/arch/arm64/boot/dts/ti/k3-j7200-evm-mcspi-loopback.dtso
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * MCSPI internal master-slave loopback example overlay for J7200.
+ *
+ * Copyright (C) 2024 Texas Instruments Incorporated - https://www.ti.com/
+ *
+ * J7200 has MCSPI4 connected as master to MCU_MCSPI2 by default and not
+ * pinned out to external pads, This overlay enables spidev on these
+ * interfaces for userspace testing.
+ */
+
+/dts-v1/;
+/plugin/;
+
+&main_spi4 {
+	status = "okay";
+	#address-cells = <0>;
+	#size-cells = <0>;
+	spi-slave;
+	dmas = <&main_udmap 0xc610>, <&main_udmap 0x4610>;
+	dma-names = "tx0", "rx0";
+
+	slave {
+		/*
+		 * Using spidev compatible is warned loudly,
+		 * thus use another equivalent compatible id
+		 * from spidev.
+		 */
+		compatible = "rohm,dh2228fv";
+		spi-max-frequency = <24000000>;
+	};
+};
+
+&mcu_spi2 {
+	status = "okay";
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	spidev@0 {
+		/*
+		 * Using spidev compatible is warned loudly,
+		 * thus use another equivalent compatible id
+		 * from spidev.
+		 */
+		compatible = "rohm,dh2228fv";
+		spi-max-frequency = <24000000>;
+		reg = <0>;
+	};
+};
-- 
2.47.2

